source/common/http/async_client_impl.h

```diff
@@ -93,7 +98,22 @@ class AsyncStreamImpl : public AsyncClient::Stream,
 public:
   AsyncStreamImpl(AsyncClientImpl& parent, AsyncClient::StreamCallbacks& callbacks,
                   const AsyncClient::StreamOptions& options);
-  ~AsyncStreamImpl() override { router_.onDestroy(); }
+  ~AsyncStreamImpl() override {
+    router_.onDestroy();
+    if (destructor_callback_.has_value()) {
+      (*destructor_callback_)();
+    }
+  }
+
+  void setDestructorCallback(AsyncClient::StreamDestructorCallbacks callback) override {
+    destructor_callback_.emplace(callback);
+  }
+  void removeDestructorCallback() override { destructor_callback_.reset(); }
+
+  void setWatermarkCallbacks(DecoderFilterWatermarkCallbacks& callbacks) override {
```

Comment:
can't remember - can we have async H2 clients, such that the downstream connection may be over the high watermark when a stream is created?  If so we need to check for high watermark and call high watermark callbacks N times when we set a new callback. and unit test.  optional bonus points for e2e.

Reply:
Yes, this can happen, but the connection only will call high/low watermark callbacks once per stream.